### PR TITLE
Performance tune the recent builds query

### PR DESCRIPTION
**Pull request by thorncp**
The recent builds query was... not great. The columns used for the
dense ranking partially covered by two indexes, but it wasn't a close
enough match for the query analyzer to use them. To keep the same query,
I think we'd want to add an index specifically for it, which would be
something like `ON (repo_id, pull_request_number, created_at DESC)`.
That seems a bit much to me, especially given that our `builds` sees a
lot of writes.

This commit completely replaces the query with a different one, making
use of `distinct on` to get the most recent build for each repo+pr. The
performance is quite a bit better, but is way more verbose.

### Before

```
Limit  (cost=1371574.44..1371574.44 rows=1 width=770) (actual time=22822.321..22822.330 rows=20 loops=1)
  ->  Sort  (cost=1371574.44..1371574.44 rows=1 width=770) (actual time=22822.321..22822.326 rows=20 loops=1)
        Sort Key: builds.id
        Sort Method: top-N heapsort  Memory: 35kB
        ->  Nested Loop  (cost=1356883.51..1371574.44 rows=1 width=770) (actual time=18168.927..22819.782 rows=3447 loops=1)
              ->  Merge Join  (cost=1356883.42..1371570.99 rows=1 width=774) (actual time=18168.886..22803.472 rows=3447 loops=1)
                    Merge Cond: (builds.repo_id = memberships.repo_id)
                    ->  Subquery Scan on builds  (cost=1356883.34..1371515.68 rows=1068 width=770) (actual time=18167.533..22697.246 rows=589733 loops=1)
                          Filter: (builds.rank = 1)
                          ->  Unique  (cost=1356883.34..1370768.05 rows=213611 width=762) (actual time=18167.521..22439.612 rows=589733 loops=1)
                                ->  WindowAgg  (cost=1356883.34..1368631.94 rows=2136109 width=762) (actual time=18167.520..21845.636 rows=2154872 loops=1)
                                      ->  Sort  (cost=1356883.34..1357951.39 rows=2136109 width=762) (actual time=18167.498..19872.591 rows=2154872 loops=1)
                                            Sort Key: builds_1.repo_id, builds_1.pull_request_number, builds_1.id
                                            Sort Method: external merge  Disk: 819640kB
                                            ->  Seq Scan on builds builds_1  (cost=0.00..279373.33 rows=2136109 width=762) (actual time=0.044..2184.130 rows=2228739 loops=1)
                    ->  Index Only Scan using index_memberships_on_user_id_and_repo_id on memberships  (cost=0.08..54.73 rows=79 width=4) (actual time=0.057..1.502 rows=3699 loops=1)
                          Index Cond: (user_id = 4434)
                          Heap Fetches: 578
              ->  Index Only Scan using repos_pkey on repos  (cost=0.08..3.44 rows=1 width=4) (actual time=0.003..0.004 rows=1 loops=3447)
                    Index Cond: (id = builds.repo_id)
                    Heap Fetches: 3447
Total runtime: 22824.363 ms
```

### After

```
 Limit  (cost=80340.95..80340.96 rows=20 width=1132) (actual time=111.735..111.742 rows=20 loops=1)
   CTE recent_builds
     ->  Nested Loop  (cost=0.26..77195.82 rows=777 width=4) (actual time=0.038..16.977 rows=13474 loops=1)
           ->  Nested Loop  (cost=0.17..366.08 rows=79 width=8) (actual time=0.020..1.378 rows=315 loops=1)
                 ->  Index Only Scan using index_memberships_on_user_id_and_repo_id on memberships  (cost=0.08..54.73 rows=79 width=4) (actual time=0.012..0.104 rows=315 loops=1)
                       Index Cond: (user_id = 4434)
                       Heap Fetches: 93
                 ->  Index Only Scan using repos_pkey on repos  (cost=0.08..3.94 rows=1 width=4) (actual time=0.003..0.003 rows=1 loops=315)
                       Index Cond: (id = memberships.repo_id)
                       Heap Fetches: 315
           ->  Index Scan using index_builds_on_repo_id on builds  (cost=0.09..970.18 rows=782 width=8) (actual time=0.003..0.036 rows=43 loops=315)
                 Index Cond: (repo_id = repos.id)
   CTE recent_builds_by_pull_request
     ->  Unique  (cost=3135.16..3136.33 rows=777 width=762) (actual time=103.512..109.036 rows=3447 loops=1)
           ->  Sort  (cost=3135.16..3135.55 rows=777 width=762) (actual time=103.511..105.319 rows=13474 loops=1)
                 Sort Key: builds_1.repo_id, builds_1.pull_request_number, builds_1.created_at, builds_1.id
                 Sort Method: quicksort  Memory: 6457kB
                 ->  Nested Loop  (cost=0.09..3127.70 rows=777 width=762) (actual time=0.050..76.240 rows=13474 loops=1)
                       ->  CTE Scan on recent_builds  (cost=0.00..4.66 rows=777 width=4) (actual time=0.039..23.591 rows=13474 loops=1)
                       ->  Index Scan using builds_pkey on builds builds_1  (cost=0.09..4.02 rows=1 width=762) (actual time=0.003..0.003 rows=1 loops=13474)
                             Index Cond: (id = recent_builds.id)
   ->  Sort  (cost=8.80..9.19 rows=777 width=1132) (actual time=111.734..111.734 rows=20 loops=1)
         Sort Key: recent_builds_by_pull_request.created_at
         Sort Method: top-N heapsort  Memory: 35kB
         ->  CTE Scan on recent_builds_by_pull_request  (cost=0.00..4.66 rows=777 width=1132) (actual time=103.514..110.976 rows=3447 loops=1)
 Total runtime: 111.965 ms
```

### CODE REVIEW

File: app/queries/recent_builds_by_repo_query.rb [hunk 1]
```diff
@@ -10,23 +10,40 @@ def initialize(user:)
...
-      builds.*,
-      dense_rank() OVER (
-        PARTITION BY repo_id, pull_request_number
-        ORDER BY created_at DESC
-      ) AS rank
+    Build.find_by_sql([<<-SQL, user_id: @user.id, limit: NUMBER_OF_BUILDS])
+      with recent_builds as (
+        select
+          builds.id
+        from
```
Comment (gylaz) [inline app/queries/recent_builds_by_repo_query.rb]: Mix of caps and lowercase keywords. Should we capitalize these?

File: app/queries/recent_builds_by_repo_query.rb [hunk 2]
```diff
@@ -10,23 +10,40 @@ def initialize(user:)
...
-
-  def ranked_by_created_at_subquery
-    Build.select(<<-SQL).to_sql
-      builds.*,
-      dense_rank() OVER (
-        PARTITION BY repo_id, pull_request_number
-        ORDER BY created_at DESC
-      ) AS rank
+    Build.find_by_sql([<<-SQL, user_id: @user.id, limit: NUMBER_OF_BUILDS])
+      with recent_builds as (
```
Comment (gylaz) [inline app/queries/recent_builds_by_repo_query.rb]: This is called `recent_builds` but I don't see anything limiting it to "recent". Should there be? Or should we rename this variable?

File: app/queries/recent_builds_by_repo_query.rb [hunk 3]
```diff
@@ -10,23 +10,40 @@ def initialize(user:)
...
+          builds.*
+        from
+          builds
+          inner join recent_builds
+            on recent_builds.id = builds.id
+        order by
+          repo_id,
+          pull_request_number,
+          created_at desc,
+          id desc
```
Comment (gylaz) [inline app/queries/recent_builds_by_repo_query.rb]: Is this line necessary for the order?

File: app/queries/recent_builds_by_repo_query.rb [hunk 4]
```diff
@@ -10,23 +10,40 @@ def initialize(user:)
...
+        select distinct on (repo_id, pull_request_number)
+          builds.*
+        from
+          builds
+          inner join recent_builds
+            on recent_builds.id = builds.id
+        order by
+          repo_id,
+          pull_request_number,
+          created_at desc,
```
Comment (gylaz) [inline app/queries/recent_builds_by_repo_query.rb]: Do we need this here if we're ordering on line 44 by the same?

Comment (thorncp) [inline app/queries/recent_builds_by_repo_query.rb]: Yep, this comes into play with the `distinct on`. It's how we get the most recent one.

File: app/queries/recent_builds_by_repo_query.rb (hunk 3 above)
Comment (thorncp) [inline app/queries/recent_builds_by_repo_query.rb]: It's the tie breaker for when two builds have the same repo_id, pull_request_number, and created_at. Not likely to happen in production but I did see it in test.

File: app/queries/recent_builds_by_repo_query.rb (hunk 2 above)
Comment (thorncp) [inline app/queries/recent_builds_by_repo_query.rb]: Good call: e6d40d4bf9655f99ac46de372eb190e0f11df75a

File: app/queries/recent_builds_by_repo_query.rb (hunk 1 above)
Comment (thorncp) [inline app/queries/recent_builds_by_repo_query.rb]: Yep. 82afcabe938425eeb26222691d231cd12760aede

**Comment by gylaz** [comment]
👍

File: app/queries/recent_builds_by_repo_query.rb [hunk 5]
```diff
@@ -10,23 +10,40 @@ def initialize(user:)
...
-
-  def ranked_by_created_at_subquery
-    Build.select(<<-SQL).to_sql
-      builds.*,
-      dense_rank() OVER (
-        PARTITION BY repo_id, pull_request_number
-        ORDER BY created_at DESC
-      ) AS rank
+    Build.find_by_sql([<<-SQL, user_id: @user.id, limit: NUMBER_OF_BUILDS])
+      WITH user_builds AS (
```
Comment (teoljungberg) [inline app/queries/recent_builds_by_repo_query.rb]: Could this be extracted into a database view instead?

**Comment by teoljungberg** [comment]
I love the idea behind this! I'd like to see this be extracted into a database view using [Scenic] though

[Scenic]: https://github.com/thoughtbot/scenic

**Comment by thorncp** [comment]
@teoljungberg I don't see the benefits of extracting a view for this, can you explain your thoughts? I wanted to be consistent with the other query classes and there's no plans for anything that would need to join/filter the view beyond what's here.

**Comment by thorncp** [comment]
Merging this so it can go out today.

**Comment by teoljungberg** [comment]
@thorncp I had an idea of moving the SQL logic out of the query classes into views or models. But, if we're fine with having those classes having that responsibility - it's a non issue.

**Pull request merged**